terraform/evaluate.go

```diff
@@ -185,6 +185,16 @@ func (d *evaluationStateData) GetForEachAttr(addr addrs.ForEachAttr, rng tfdiags
 		returnVal = d.InstanceKeyData.EachKey
 	case "value":
 		returnVal = d.InstanceKeyData.EachValue
+
+		if returnVal == cty.NilVal {
+			diags = diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  `each.value is unknown and cannot be used in this context`,
+				Detail:   fmt.Sprintf(`A reference to "each.value" has been used in a context in which it unavailable, such as after the config no longer contains the value in its "for_each" expression. Remove this reference to each.value in your config to work around this error.`),
```

Comment:
This phrasing is awkward, but I haven't been able to think of an alternative - it's an awkward situation. The only concrete suggestion I might suggest is spelling out "configuration" instead of config. I'll leave that choice to you and approve this change, no worries if you disagree.